real_time/add-custom-dynamic-entity-data-source/src/main/java/com/esri/samples/add_custom_dynamic_entity_data_source/SimulatedDataSource.java

```diff
@@ -0,0 +1,177 @@
+/*
+ * Copyright 2023 Esri.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package com.esri.samples.add_custom_dynamic_entity_data_source;
+
+import java.io.BufferedReader;
+import java.io.FileReader;
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Objects;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.CompletionException;
+import java.util.concurrent.Executors;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.ScheduledFuture;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicInteger;
+
+import com.google.gson.Gson;
+
+import com.esri.arcgisruntime.data.Field;
+import com.esri.arcgisruntime.geometry.Point;
+import com.esri.arcgisruntime.geometry.SpatialReferences;
+import com.esri.arcgisruntime.realtime.DynamicEntityDataSource;
+import com.esri.arcgisruntime.realtime.DynamicEntityDataSourceInfo;
+
+/**
+ * A custom DynamicEntityDataSource for processing observations read from a given JSON file.
+ */
+class SimulatedDataSource extends DynamicEntityDataSource {
+
+  private final String fileName;
+  private final String entityIdFieldName;
+  private final long delay;
+  private final ScheduledExecutorService executorService = Executors.newSingleThreadScheduledExecutor();
+  private ScheduledFuture<?> observationProcessing;
+  private final AtomicInteger currentLineNumber = new AtomicInteger();
+  private final List<String> allJsonLines = new ArrayList<>();
+  private final Gson gson = new Gson();
+
+  /**
+   * Construct a custom DynamicEntityDataSource.
+   *
+   * @param fileName name of a file with JSON observation data
+   * @param entityIdFieldName the name of the field containing values that uniquely identifies each entity
+   * @param delay millisecond delay between observations being added to the data source
+   */
+  SimulatedDataSource(String fileName, String entityIdFieldName, long delay) {
+    this.fileName = fileName;
+    this.entityIdFieldName = entityIdFieldName;
+    this.delay = delay;
+  }
+
+  @Override
+  protected CompletableFuture<Void> onConnectAsync() {
+    return CompletableFuture.completedFuture((Void)null).whenComplete((unused, throwable) -> {
+      if (throwable != null) {
+        System.err.println("Failed to connect custom data source: " + throwable.getMessage());
```

Comment:
I think the designed way to handle this would be to look at the connection status and error when `connectAsync` completes - probably use the properties on `DynamicEntityDataSource`.

Here is where you might check if a cancellation exception occurred because that would mean disconnect had been called during connecting and you _might_ need to do some clean up etc. because of that.

Reply:
I think we should use the `completeAsync` variant here (see https://devtopia.esri.com/runtime/java/blob/main/arcgis-java-testFX/src/test/java/com/esri/arcgisruntime/testfx/realtime/CustomDynamicEntityDataSourceTest.java#L99) so that the loading up of the data is async.